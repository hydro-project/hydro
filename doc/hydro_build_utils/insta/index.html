<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `insta` mod in crate `hydro_build_utils`."><title>hydro_build_utils::insta - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="hydro_build_utils" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (f6092f224 2025-09-22)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module insta</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../hydro_build_utils/index.html">hydro_<wbr>build_<wbr>utils</a><span class="version">0.0.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module insta</a></h2><h3><a href="#modules">Module Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate hydro_<wbr>build_<wbr>utils</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">hydro_build_utils</a></div><h1>Module <span>insta</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/hydro_build_utils/lib.rs.html#20">Source</a> </span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="internals/index.html" title="mod hydro_build_utils::insta::internals">internals</a></dt><dd>Exposes some library internals.</dd></dl><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="macro" href="macro.allow_duplicates.html" title="macro hydro_build_utils::insta::allow_duplicates">allow_<wbr>duplicates</a></dt><dd>Utility macro to permit a multi-snapshot run where all snapshots match.</dd><dt><a class="macro" href="macro.assert_binary_snapshot.html" title="macro hydro_build_utils::insta::assert_binary_snapshot">assert_<wbr>binary_<wbr>snapshot</a></dt><dd>(Experimental)
Asserts a binary snapshot in the form of a <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec&lt;u8&gt;</code></a>.</dd><dt><a class="macro" href="macro.assert_compact_debug_snapshot.html" title="macro hydro_build_utils::insta::assert_compact_debug_snapshot">assert_<wbr>compact_<wbr>debug_<wbr>snapshot</a></dt><dd>Asserts a <a href="https://doc.rust-lang.org/nightly/core/fmt/macros/derive.Debug.html" title="derive core::fmt::macros::Debug"><code>Debug</code></a> snapshot in compact format.</dd><dt><a class="macro" href="macro.assert_debug_snapshot.html" title="macro hydro_build_utils::insta::assert_debug_snapshot">assert_<wbr>debug_<wbr>snapshot</a></dt><dd>Asserts a <a href="https://doc.rust-lang.org/nightly/core/fmt/macros/derive.Debug.html" title="derive core::fmt::macros::Debug"><code>Debug</code></a> snapshot.</dd><dt><a class="macro" href="macro.assert_display_snapshot.html" title="macro hydro_build_utils::insta::assert_display_snapshot">assert_<wbr>display_<wbr>snapshot</a><wbr><span class="stab deprecated" title="">Deprecated</span></dt><dd>Asserts a <a href="https://doc.rust-lang.org/nightly/core/fmt/trait.Display.html" title="trait core::fmt::Display"><code>Display</code></a> snapshot.</dd><dt><a class="macro" href="macro.assert_snapshot.html" title="macro hydro_build_utils::insta::assert_snapshot">assert_<wbr>snapshot</a></dt><dd>Asserts a <a href="https://doc.rust-lang.org/nightly/alloc/string/struct.String.html" title="struct alloc::string::String"><code>String</code></a> snapshot.</dd><dt><a class="macro" href="macro.elog.html" title="macro hydro_build_utils::insta::elog">elog</a></dt><dt><a class="macro" href="macro.with_settings.html" title="macro hydro_build_utils::insta::with_settings">with_<wbr>settings</a></dt><dd>Settings configuration macro.</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.MetaData.html" title="struct hydro_build_utils::insta::MetaData">Meta<wbr>Data</a></dt><dd>Snapshot metadata information.</dd><dt><a class="struct" href="struct.Settings.html" title="struct hydro_build_utils::insta::Settings">Settings</a></dt><dd>Configures how insta operates at test time.</dd><dt><a class="struct" href="struct.Snapshot.html" title="struct hydro_build_utils::insta::Snapshot">Snapshot</a></dt><dd>A helper to work with file snapshots.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="enum" href="enum.TextSnapshotKind.html" title="enum hydro_build_utils::insta::TextSnapshotKind">Text<wbr>Snapshot<wbr>Kind</a></dt></dl></section></div></main></body></html>