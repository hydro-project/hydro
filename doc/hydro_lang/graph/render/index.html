<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `render` mod in crate `hydro_lang`."><title>hydro_lang::graph::render - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="hydro_lang" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (9f32ccf35 2025-09-21)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module render</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../hydro_lang/index.html">hydro_<wbr>lang</a><span class="version">0.14.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module render</a></h2><h3><a href="#reexports">Module Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In hydro_<wbr>lang::<wbr>graph</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">hydro_lang</a>::<wbr><a href="../index.html">graph</a></div><h1>Module <span>render</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/hydro_lang/graph/render.rs.html#1-1096">Source</a> </span></div><span class="item-info"><div class="stab portability">Available on <strong>crate feature <code>viz</code></strong> only.</div></span><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><dl class="item-table reexports"><dt id="reexport.HydroDot"><code>pub use super::graphviz::<a class="struct" href="../graphviz/struct.HydroDot.html" title="struct hydro_lang::graph::graphviz::HydroDot">HydroDot</a>;</code></dt><dt id="reexport.escape_dot"><code>pub use super::graphviz::<a class="fn" href="../graphviz/fn.escape_dot.html" title="fn hydro_lang::graph::graphviz::escape_dot">escape_dot</a>;</code></dt><dt id="reexport.HydroMermaid"><code>pub use super::mermaid::<a class="struct" href="../mermaid/struct.HydroMermaid.html" title="struct hydro_lang::graph::mermaid::HydroMermaid">HydroMermaid</a>;</code></dt><dt id="reexport.escape_mermaid"><code>pub use super::mermaid::<a class="fn" href="../mermaid/fn.escape_mermaid.html" title="fn hydro_lang::graph::mermaid::escape_mermaid">escape_mermaid</a>;</code></dt><dt id="reexport.HydroReactFlow"><code>pub use super::reactflow::<a class="struct" href="../reactflow/struct.HydroReactFlow.html" title="struct hydro_lang::graph::reactflow::HydroReactFlow">HydroReactFlow</a>;</code></dt></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.HydroGraphStructure.html" title="struct hydro_lang::graph::render::HydroGraphStructure">Hydro<wbr>Graph<wbr>Structure</a></dt><dd>Graph structure tracker for Hydro IR rendering.</dd><dt><a class="struct" href="struct.HydroWriteConfig.html" title="struct hydro_lang::graph::render::HydroWriteConfig">Hydro<wbr>Write<wbr>Config</a></dt><dd>Configuration for graph writing.</dd><dt><a class="struct" href="struct.IndentedGraphWriter.html" title="struct hydro_lang::graph::render::IndentedGraphWriter">Indented<wbr>Graph<wbr>Writer</a></dt><dd>Base struct for text-based graph writers that use indentation.
Contains common fields shared by DOT and Mermaid writers.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.HydroEdgeType.html" title="enum hydro_lang::graph::render::HydroEdgeType">Hydro<wbr>Edge<wbr>Type</a></dt><dd>Types of edges in Hydro IR.</dd><dt><a class="enum" href="enum.HydroNodeType.html" title="enum hydro_lang::graph::render::HydroNodeType">Hydro<wbr>Node<wbr>Type</a></dt><dd>Types of nodes in Hydro IR for styling purposes.</dd><dt><a class="enum" href="enum.NodeLabel.html" title="enum hydro_lang::graph::render::NodeLabel">Node<wbr>Label</a></dt><dd>Label for a graph node - can be either a static string or contain expressions.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.HydroGraphWrite.html" title="trait hydro_lang::graph::render::HydroGraphWrite">Hydro<wbr>Graph<wbr>Write</a></dt><dd>Trait for writing textual representations of Hydro IR graphs, i.e. mermaid or dot graphs.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.extract_op_name.html" title="fn hydro_lang::graph::render::extract_op_name">extract_<wbr>op_<wbr>name</a></dt><dd>Function to extract an op_name from a print_root() result for use in labels.</dd><dt><a class="fn" href="fn.extract_short_label.html" title="fn hydro_lang::graph::render::extract_short_label">extract_<wbr>short_<wbr>label</a></dt><dd>Extract a short, readable label from the full token stream label using print_root() style naming</dd><dt><a class="fn" href="fn.render_hydro_ir_dot.html" title="fn hydro_lang::graph::render::render_hydro_ir_dot">render_<wbr>hydro_<wbr>ir_<wbr>dot</a></dt><dt><a class="fn" href="fn.render_hydro_ir_mermaid.html" title="fn hydro_lang::graph::render::render_hydro_ir_mermaid">render_<wbr>hydro_<wbr>ir_<wbr>mermaid</a></dt><dt><a class="fn" href="fn.render_hydro_ir_reactflow.html" title="fn hydro_lang::graph::render::render_hydro_ir_reactflow">render_<wbr>hydro_<wbr>ir_<wbr>reactflow</a></dt><dt><a class="fn" href="fn.write_hydro_ir_dot.html" title="fn hydro_lang::graph::render::write_hydro_ir_dot">write_<wbr>hydro_<wbr>ir_<wbr>dot</a></dt><dt><a class="fn" href="fn.write_hydro_ir_mermaid.html" title="fn hydro_lang::graph::render::write_hydro_ir_mermaid">write_<wbr>hydro_<wbr>ir_<wbr>mermaid</a></dt><dt><a class="fn" href="fn.write_hydro_ir_reactflow.html" title="fn hydro_lang::graph::render::write_hydro_ir_reactflow">write_<wbr>hydro_<wbr>ir_<wbr>reactflow</a></dt></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><dl class="item-table"><dt><a class="type" href="type.GraphWriteError.html" title="type hydro_lang::graph::render::GraphWriteError">Graph<wbr>Write<wbr>Error</a></dt><dd>Common error type used by all graph writers.</dd></dl></section></div></main></body></html>