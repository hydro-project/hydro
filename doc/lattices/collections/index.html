<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Simple singleton or array collection with `cc_traits` implementations."><title>lattices::collections - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module collections</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate lattices</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a></div><h1>Module <span>collections</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/collections.rs.html#1-1203">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Simple singleton or array collection with [<code>cc_traits</code>] implementations.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">Array<wbr>Map</a></dt><dd>An array wrapper representing a fixed-size map.</dd><dt><a class="struct" href="struct.ArraySet.html" title="struct lattices::collections::ArraySet">Array<wbr>Set</a></dt><dd>An array wrapper representing a fixed-size set (modulo duplicate items).</dd><dt><a class="struct" href="struct.EmptyMap.html" title="struct lattices::collections::EmptyMap">Empty<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.EmptySet.html" title="struct lattices::collections::EmptySet">Empty<wbr>Set</a></dt><dd>A type that will always be an empty set.</dd><dt><a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">Option<wbr>Map</a></dt><dd>A key-value entry wrapper around <code>Option&lt;(K, V)&gt;</code> representing a singleton or empty map.</dd><dt><a class="struct" href="struct.OptionSet.html" title="struct lattices::collections::OptionSet">Option<wbr>Set</a></dt><dd>A wrapper around <code>Option</code>, representing either a singleton or empty set.</dd><dt><a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">Singleton<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.SingletonSet.html" title="struct lattices::collections::SingletonSet">Singleton<wbr>Set</a></dt><dd>A wrapper around an item, representing a singleton set.</dd><dt><a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively implemented map.</dd><dt><a class="struct" href="struct.VecSet.html" title="struct lattices::collections::VecSet">VecSet</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively-implemented set.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMap<wbr>Values</a></dt><dd>Trait for transforming the values of a map without changing the overall type of the data structure.</dd></dl></section></div></main></body></html>