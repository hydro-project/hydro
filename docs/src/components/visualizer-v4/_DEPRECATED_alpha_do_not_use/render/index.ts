/**
 * @fileoverview Render module exports
 * 
 * Central export point for ReactFlow rendering functionality.
 * 
 * @example
 * ```typescript
 * import { FlowGraph } from './vis/render';
 * import './vis/render/styles.css'; // Import styles
 * ```
 */

export { FlowGraph } from './FlowGraph';
export { ReactFlowConverter } from './ReactFlowConverter';
export { GraphStandardNode, GraphContainerNode } from './nodes';
export { GraphStandardEdge, GraphHyperEdge } from './edges';
export { 
  createNodeEventHandlers, 
  createEdgeEventHandlers, 
  createContainerEventHandlers 
} from './eventHandlers';
export { 
  DEFAULT_RENDER_CONFIG
} from './config';
export type {
  RenderConfig,
  FlowGraphEventHandlers
} from './types';
