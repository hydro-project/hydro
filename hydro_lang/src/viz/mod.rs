//! Graph visualization utilities for Hydro IR

pub mod api;
pub mod config;
pub mod debug;
pub mod graphviz;
pub mod json;
pub mod mermaid;
pub mod render;

#[cfg(test)]
mod json_test;

#[cfg(test)]
mod url_generation_test;

// Re-export only the necessary public API
pub use api::GraphApi;
pub use config::VisualizerConfig;
