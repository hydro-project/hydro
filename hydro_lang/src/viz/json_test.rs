//! Tests for JSON graph generation with semantic tags

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    #[cfg(test)]
    use hydro_build_utils::insta;

    use crate::viz::json::HydroJson;
    use crate::viz::render::{
        HydroEdgeProp, HydroGraphWrite, HydroNodeType, HydroWriteConfig, NodeLabel,
    };

    #[test]
    fn test_json_structure_with_semantic_tags() {
        let mut output = String::new();
        let config = HydroWriteConfig::default();
        let mut writer = HydroJson::new(&mut output, &config);

        // Write a simple graph
        writer.write_prologue().unwrap();

        // Add a source node
        writer
            .write_node_definition(
                0,
                &NodeLabel::Static("source".to_string()),
                HydroNodeType::Source,
                Some(0),
                Some("Process"),
                None,
            )
            .unwrap();

        // Add a transform node
        writer
            .write_node_definition(
                1,
                &NodeLabel::Static("map".to_string()),
                HydroNodeType::Transform,
                Some(0),
                Some("Process"),
                None,
            )
            .unwrap();

        // Add an edge with semantic properties
        let mut edge_props = HashSet::new();
        edge_props.insert(HydroEdgeProp::Stream);
        edge_props.insert(HydroEdgeProp::Unbounded);
        edge_props.insert(HydroEdgeProp::TotalOrder);

        writer.write_edge(0, 1, &edge_props, None).unwrap();

        writer.write_epilogue().unwrap();

        // Snapshot test the complete JSON output
        insta::assert_snapshot!(output);
    }

    #[test]
    fn test_empty_semantic_tags() {
        let mut output = String::new();
        let config = HydroWriteConfig::default();
        let mut writer = HydroJson::new(&mut output, &config);

        writer.write_prologue().unwrap();

        writer
            .write_node_definition(
                0,
                &NodeLabel::Static("node".to_string()),
                HydroNodeType::Transform,
                None,
                None,
                None,
            )
            .unwrap();

        // Edge with no properties
        let edge_props = HashSet::new();
        writer.write_edge(0, 0, &edge_props, None).unwrap();

        writer.write_epilogue().unwrap();

        // Snapshot test the complete JSON output
        insta::assert_snapshot!(output);
    }

    #[test]
    fn test_deterministic_output_and_network_tagging() {
        use crate::viz::render::HydroWriteConfig;
        let mut output1 = String::new();
        let mut output2 = String::new();
        let config = HydroWriteConfig::default();
        let mut w1 = HydroJson::new(&mut output1, &config);
        let mut w2 = HydroJson::new(&mut output2, &config);

        // Build same small graph with two locations to force Network tag
        let mut edge_props = HashSet::new();
        edge_props.insert(HydroEdgeProp::Stream);

        // Graph 1
        w1.write_prologue().unwrap();
        w1.write_node_definition(
            0,
            &NodeLabel::Static("a".into()),
            HydroNodeType::Source,
            Some(0),
            Some("Process"),
            None,
        )
        .unwrap();
        w1.write_node_definition(
            1,
            &NodeLabel::Static("b".into()),
            HydroNodeType::Transform,
            Some(1),
            Some("Process"),
            None,
        )
        .unwrap();
        w1.write_edge(0, 1, &edge_props, None).unwrap();
        w1.write_epilogue().unwrap();

        // Graph 2 (same operations, different insertion order to test determinism)
        w2.write_prologue().unwrap();
        w2.write_node_definition(
            1,
            &NodeLabel::Static("b".into()),
            HydroNodeType::Transform,
            Some(1),
            Some("Process"),
            None,
        )
        .unwrap();
        w2.write_node_definition(
            0,
            &NodeLabel::Static("a".into()),
            HydroNodeType::Source,
            Some(0),
            Some("Process"),
            None,
        )
        .unwrap();
        w2.write_edge(0, 1, &edge_props, None).unwrap();
        w2.write_epilogue().unwrap();

        // Verify deterministic output
        assert_eq!(
            output1, output2,
            "JSON output should be deterministic regardless of insertion order"
        );

        // Snapshot test the complete JSON output
        insta::assert_snapshot!(output1);
    }
}
