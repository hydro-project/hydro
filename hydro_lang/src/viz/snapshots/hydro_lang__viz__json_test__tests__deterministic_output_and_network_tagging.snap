---
source: hydro_lang/src/viz/json_test.rs
expression: output1
---
{
  "nodes": [
    {
      "data": {
        "backtrace": [],
        "locationId": 0,
        "locationType": "Process"
      },
      "fullLabel": "a [hydro operator]",
      "id": "0",
      "label": "a",
      "nodeType": "Source",
      "shortLabel": "a"
    },
    {
      "data": {
        "backtrace": [],
        "locationId": 1,
        "locationType": "Process"
      },
      "fullLabel": "b [hydro operator]",
      "id": "1",
      "label": "b",
      "nodeType": "Transform",
      "shortLabel": "b"
    }
  ],
  "edges": [
    {
      "id": "e0",
      "semanticTags": [
        "Network",
        "Stream"
      ],
      "source": "0",
      "target": "1"
    }
  ],
  "hierarchyChoices": [
    {
      "children": [],
      "id": "location",
      "name": "Location"
    }
  ],
  "nodeAssignments": {
    "location": {
      "0": "loc_0",
      "1": "loc_1"
    }
  },
  "selectedHierarchy": "location",
  "edgeStyleConfig": {
    "note": "Edge styles are now computed per-edge using the unified edge style system. This config is provided for reference and compatibility.",
    "semanticMappings": {
      "BoundednessGroup": {
        "Bounded": {
          "halo": "none"
        },
        "Unbounded": {
          "halo": "light-blue"
        }
      },
      "CollectionGroup": {
        "Optional": {
          "arrowhead": "diamond-open",
          "color": "#6b7280"
        },
        "Singleton": {
          "arrowhead": "circle-filled",
          "color": "#000000"
        },
        "Stream": {
          "arrowhead": "triangle-filled",
          "color": "#2563eb"
        }
      },
      "KeyednessGroup": {
        "Keyed": {
          "line-style": "hash-marks"
        },
        "NotKeyed": {
          "line-style": "single"
        }
      },
      "NetworkGroup": {
        "Local": {
          "animation": "static",
          "line-pattern": "solid"
        },
        "Network": {
          "animation": "animated",
          "line-pattern": "dashed"
        }
      },
      "OrderingGroup": {
        "NoOrder": {
          "waviness": "wavy"
        },
        "TotalOrder": {
          "waviness": "straight"
        }
      }
    },
    "semanticPriorities": [
      [
        "Unbounded",
        "Bounded"
      ],
      [
        "NoOrder",
        "TotalOrder"
      ],
      [
        "Keyed",
        "NotKeyed"
      ],
      [
        "Network",
        "Local"
      ]
    ]
  },
  "nodeTypeConfig": {
    "defaultType": "Transform",
    "types": [
      {
        "colorIndex": 0,
        "id": "Aggregation",
        "label": "Aggregation"
      },
      {
        "colorIndex": 1,
        "id": "Join",
        "label": "Join"
      },
      {
        "colorIndex": 2,
        "id": "Network",
        "label": "Network"
      },
      {
        "colorIndex": 3,
        "id": "Sink",
        "label": "Sink"
      },
      {
        "colorIndex": 4,
        "id": "Source",
        "label": "Source"
      },
      {
        "colorIndex": 5,
        "id": "Tee",
        "label": "Tee"
      },
      {
        "colorIndex": 6,
        "id": "Transform",
        "label": "Transform"
      }
    ]
  },
  "legend": {
    "items": [
      {
        "label": "Aggregation",
        "type": "Aggregation"
      },
      {
        "label": "Join",
        "type": "Join"
      },
      {
        "label": "Network",
        "type": "Network"
      },
      {
        "label": "Sink",
        "type": "Sink"
      },
      {
        "label": "Source",
        "type": "Source"
      },
      {
        "label": "Tee",
        "type": "Tee"
      },
      {
        "label": "Transform",
        "type": "Transform"
      }
    ],
    "title": "Node Types"
  }
}
